src/components/shared/ReactFlow/FlowCanvas/FlowCanvas.tsx

```diff
@@ -366,58 +339,58 @@ const FlowCanvas = ({
   );
 
   const onConnectEnd = useCallback(
-    (_e: MouseEvent | TouchEvent, connectionState: FinalConnectionState) => {
+    (e: MouseEvent | TouchEvent, connectionState: FinalConnectionState) => {
       if (connectionState.isValid) {
         // Valid connections are handled by onConnect
         return;
       }
 
-      const ghostNode = reactFlowInstance
-        ?.getNodes()
-        .find((node) => node.type === "ghost");
+      const position = latestFlowPosRef.current;
+      if (!position) return;
 
-      if (!ghostNode) {
+      if (!(e instanceof MouseEvent) || !(e.metaKey || e.ctrlKey)) {
         return;
       }
 
-      const { componentRef } = ghostNode.data as {
-        componentRef: ComponentReference;
-      };
+      const { fromHandle } = connectionState;
+      if (!fromHandle || !fromHandle.nodeId || !fromHandle.id) return;
 
-      const position = latestFlowPosRef.current;
-      if (!position) return;
+      if (!isGraphImplementation(currentSubgraphSpec.implementation)) return;
 
-      let newComponentSpec = { ...componentSpec };
-      const fromHandle = connectionState.fromHandle;
+      // Only allow creating IO nodes when dragging from task nodes
+      // Input/output nodes cannot be directly connected to other input/output nodes
+      if (!fromHandle.nodeId.startsWith("task_")) {
```

Comment:
I don't like this kind of hard-coded check. What if a task somehow doesn't have the `task_` prefix for its id? I'd prefer to see us fetch the node via id and then check against it's type.

e.g. something like:

```suggestion
     const FAST_PLACE_NODES = new Set(["task"])

     ...      

     const fromNode = nodes.find(n => n.id = fromHandle.nodeId)
      if (!FAST_PLACE_NODES.includes(fromNode.type)) {
```

Also, side note: this is the kind of reactflow operation that the node manager is intended to simplify.